### ISSUE TITLE
Allow arbitrary id names in augment.mlogit

### ISSUE THREAD

**Issue by gregmacfarlane** [tidymodels/broom#1157]
## The problem

`mlogit` relies on indexed data frames in the `dfidx` package --- each row of the data has to be assigned to an observation and an alternative using a two-column index.  As currently implemented, the `augment.mlogit` method uses hard-coded variable names to identify the observation / alternative pairs. This means that if the user has data without these names, `augment` will error out. These variable names are the default --- `dfidx` uses index names of `id1` and `id2` when transforming wide data --- but users might easily have other names because their data is already in long format, or because they used the `idnames` argument to `dfidx`.

This needs to be changed to allow for flexible names. The mlogit package *always* uses the first index for the observation and the second for the alternative, so using a column location would be safe.

## Reproducible example

``` r
library(mlogit)
#> Loading required package: dfidx
#> 
#> Attaching package: 'dfidx'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(broom)

# The Fishing dataset from the mlogit package is in a "wide" 
# setup, with one observation per row.
data("Fishing", package = "mlogit")
head(Fishing)
#>      mode price.beach price.pier price.boat price.charter catch.beach
#> 1 charter     157.930    157.930    157.930       182.930      0.0678
#> 2 charter      15.114     15.114     10.534        34.534      0.1049
#> 3    boat     161.874    161.874     24.334        59.334      0.5333
#> 4    pier      15.134     15.134     55.930        84.930      0.0678
#> 5    boat     106.930    106.930     41.514        71.014      0.0678
#> 6 charter     192.474    192.474     28.934        63.934      0.5333
#>   catch.pier catch.boat catch.charter   income
#> 1     0.0503     0.2601        0.5391 7083.332
#> 2     0.0451     0.1574        0.4671 1250.000
#> 3     0.4522     0.2413        1.0266 3750.000
#> 4     0.0789     0.1643        0.5391 2083.333
#> 5     0.0503     0.1082        0.3240 4583.332
#> 6     0.4522     0.1665        0.3975 4583.332
# to be used in mlogit, needs to be in "long", with one row representing
# an observation-choice pair.
Fish <- dfidx(Fishing, varying = 2:9, shape = "wide", choice = "mode")
# dfidx uses index names of `id1` and `id2` when transforming wide data
# into long data by default, but the `idnames` argument can allow for changing.
head(Fish)
#> ~~~~~~~
#>  first 10 observations out of 4728 
#> ~~~~~~~
#>     mode   income   price  catch    idx
#> 1  FALSE 7083.332 157.930 0.0678 1:each
#> 2  FALSE 7083.332 157.930 0.2601 1:boat
#> 3   TRUE 7083.332 182.930 0.5391 1:rter
#> 4  FALSE 7083.332 157.930 0.0503 1:pier
#> 5  FALSE 1250.000  15.114 0.1049 2:each
#> 6  FALSE 1250.000  10.534 0.1574 2:boat
#> 7   TRUE 1250.000  34.534 0.4671 2:rter
#> 8  FALSE 1250.000  15.114 0.0451 2:pier
#> 9  FALSE 3750.000 161.874 0.5333 3:each
#> 10  TRUE 3750.000  24.334 0.2413 3:boat
#> 
#> ~~~ indexes ~~~~
#>    id1     id2
#> 1    1   beach
#> 2    1    boat
#> 3    1 charter
#> 4    1    pier
#> 5    2   beach
#> 6    2    boat
#> 7    2 charter
#> 8    2    pier
#> 9    3   beach
#> 10   3    boat
#> indexes:  1, 2
m1 <- mlogit(mode ~ price + catch, data = Fish)
augment(m1)
#> # A tibble: 4,728 × 8
#>       id alternative chosen price  catch .probability .fitted .resid
#>    <int> <fct>       <lgl>  <dbl>  <dbl>        <dbl>   <dbl>  <dbl>
#>  1     1 beach       FALSE  158.  0.0678      0.128   -3.89   -0.369
#>  2     1 boat        FALSE  158.  0.260       0.330   -2.95   -0.385
#>  3     1 charter     TRUE   183.  0.539       0.369   -2.83   -0.478
#>  4     1 pier        FALSE  158.  0.0503      0.173   -3.59   -0.324
#>  5     2 beach       FALSE   15.1 0.105       0.121   -0.335  -0.461
#>  6     2 boat        FALSE   10.5 0.157       0.332    0.670  -0.457
#>  7     2 charter     TRUE    34.5 0.467       0.385    0.819   0.604
#>  8     2 pier        FALSE   15.1 0.0451      0.161   -0.0506 -0.265
#>  9     3 beach       FALSE  162.  0.533       0.00738 -3.81   -0.309
#> 10     3 boat        TRUE    24.3 0.241       0.478    0.359  -0.309
#> # ℹ 4,718 more rows

# if the data are already in long format, mlogit will default to using the
# first two columns as the indices. In this case, the names of the columns
# are different and augment fails. If we use the TravelMode data also used
# in the mlogit examples, the index columns are names "individual" and "mode".
data("TravelMode", package = "AER")
m2 <- mlogit(choice ~ wait + travel + vcost, TravelMode, heterosc = TRUE)
head(m2$model)
#> ~~~~~~~
#>  first 10 observations out of 840 
#> ~~~~~~~
#>    choice wait travel vcost    idx probabilities   linpred
#> 1   FALSE   69    100    59  1:air    0.09368642 -4.088866
#> 2   FALSE   34    372    31 1:rain    0.32229408 -2.594278
#> 3   FALSE   35    417    25  1:bus    0.07968064 -2.789754
#> 4    TRUE    0    180    10  1:car    0.49881521 -2.242450
#> 5   FALSE   64     68    58  2:air    0.17581996 -3.758902
#> 6   FALSE   44    354    31 2:rain    0.26345967 -2.992226
#> 7   FALSE   53    399    25  2:bus    0.01369371 -3.547953
#> 8    TRUE    0    255    11  2:car    0.54439205 -2.472358
#> 9   FALSE   69    125   115  3:air    0.26178393 -4.817235
#> 10  FALSE   34    892    98 3:rain    0.13660546 -4.891552
#> 
#> ~~~ indexes ~~~~
#>    individual  mode
#> 1           1   air
#> 2           1 train
#> 3           1   bus
#> 4           1   car
#> 5           2   air
#> 6           2 train
#> 7           2   bus
#> 8           2   car
#> 9           3   air
#> 10          3 train
#> indexes:  1, 2
augment(m2)
#> Error in `dplyr::select()`:
#> ! Can't subset columns that don't exist.
#> ✖ Column `id` doesn't exist.
#> Backtrace:
#>      ▆
#>   1. ├─generics::augment(m2)
#>   2. ├─broom:::augment.mlogit(m2)
#>   3. │ └─... %>% dplyr::select(id, alternative, chosen, everything())
#>   4. ├─dplyr::select(., id, alternative, chosen, everything())
#>   5. ├─dplyr:::select.data.frame(., id, alternative, chosen, everything())
#>   6. │ └─tidyselect::eval_select(expr(c(...)), data = .data, error_call = error_call)
#>   7. │   └─tidyselect:::eval_select_impl(...)
#>   8. │     ├─tidyselect:::with_subscript_errors(...)
#>   9. │     │ └─rlang::try_fetch(...)
#>  10. │     │   └─base::withCallingHandlers(...)
#>  11. │     └─tidyselect:::vars_select_eval(...)
#>  12. │       └─tidyselect:::walk_data_tree(expr, data_mask, context_mask)
#>  13. │         └─tidyselect:::eval_c(expr, data_mask, context_mask)
#>  14. │           └─tidyselect:::reduce_sels(node, data_mask, context_mask, init = init)
#>  15. │             └─tidyselect:::walk_data_tree(new, data_mask, context_mask)
#>  16. │               └─tidyselect:::as_indices_sel_impl(...)
#>  17. │                 └─tidyselect:::as_indices_impl(...)
#>  18. │                   └─tidyselect:::chr_as_locations(x, vars, call = call, arg = arg)
#>  19. │                     └─vctrs::vec_as_location(...)
#>  20. └─vctrs (local) `<fn>`()
#>  21.   └─vctrs:::stop_subscript_oob(...)
#>  22.     └─vctrs:::stop_subscript(...)
#>  23.       └─rlang::abort(...)

```

## Session Info
As a note, I am using my own local version of mlogit, but my changes are not material to this issue.
```
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.0 (2022-04-22)
#>  os       macOS 13.3.1
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/Denver
#>  date     2023-05-18
#>  pandoc   2.19.2 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version  date (UTC) lib source
#>  backports     1.4.1    2021-12-13 [1] CRAN (R 4.2.0)
#>  broom       * 1.0.1    2022-08-29 [1] CRAN (R 4.2.0)
#>  cli           3.6.1    2023-03-23 [1] CRAN (R 4.2.0)
#>  dfidx       * 0.0-5    2022-06-28 [1] CRAN (R 4.2.0)
#>  digest        0.6.29   2021-12-01 [1] CRAN (R 4.2.0)
#>  dplyr         1.1.2    2023-04-20 [1] CRAN (R 4.2.0)
#>  evaluate      0.15     2022-02-18 [1] CRAN (R 4.2.0)
#>  fansi         1.0.4    2023-01-22 [1] CRAN (R 4.2.0)
#>  fastmap       1.1.0    2021-01-25 [1] CRAN (R 4.2.0)
#>  Formula       1.2-5    2023-02-24 [1] CRAN (R 4.2.0)
#>  fs            1.5.2    2021-12-08 [1] CRAN (R 4.2.0)
#>  generics      0.1.3    2022-07-05 [1] CRAN (R 4.2.0)
#>  glue          1.6.2    2022-02-24 [1] CRAN (R 4.2.0)
#>  highr         0.9      2021-04-16 [1] CRAN (R 4.2.0)
#>  htmltools     0.5.2    2021-08-25 [1] CRAN (R 4.2.0)
#>  knitr         1.39     2022-04-26 [1] CRAN (R 4.2.0)
#>  lattice       0.20-45  2021-09-22 [1] CRAN (R 4.2.0)
#>  lifecycle     1.0.3    2022-10-07 [1] CRAN (R 4.2.0)
#>  lmtest        0.9-40   2022-03-21 [1] CRAN (R 4.2.0)
#>  magrittr      2.0.3    2022-03-30 [1] CRAN (R 4.2.0)
#>  MASS          7.3-58.1 2022-08-03 [1] CRAN (R 4.2.0)
#>  mlogit      * 1.1-2    2023-05-17 [1] Github (gregmacfarlane/mlogit@c4fe042)
#>  pillar        1.9.0    2023-03-22 [1] CRAN (R 4.2.0)
#>  pkgconfig     2.0.3    2019-09-22 [1] CRAN (R 4.2.0)
#>  purrr         0.3.5    2022-10-06 [1] CRAN (R 4.2.0)
#>  R6            2.5.1    2021-08-19 [1] CRAN (R 4.2.0)
#>  rbibutils     2.2.13   2023-01-13 [1] CRAN (R 4.2.0)
#>  Rdpack        2.4      2022-07-20 [1] CRAN (R 4.2.0)
#>  reprex        2.0.2    2022-08-17 [1] CRAN (R 4.2.0)
#>  rlang         1.1.1    2023-04-28 [1] CRAN (R 4.2.0)
#>  rmarkdown     2.14     2022-04-25 [1] CRAN (R 4.2.0)
#>  rstudioapi    0.13     2020-11-12 [1] CRAN (R 4.2.0)
#>  sessioninfo   1.2.2    2021-12-06 [1] CRAN (R 4.2.0)
#>  statmod       1.5.0    2023-01-06 [1] CRAN (R 4.2.0)
#>  stringi       1.7.6    2021-11-29 [1] CRAN (R 4.2.0)
#>  stringr       1.4.1    2022-08-20 [1] CRAN (R 4.2.0)
#>  tibble        3.2.1    2023-03-20 [1] CRAN (R 4.2.0)
#>  tidyr         1.2.1    2022-09-08 [1] CRAN (R 4.2.0)
#>  tidyselect    1.2.0    2022-10-10 [1] CRAN (R 4.2.0)
#>  utf8          1.2.3    2023-01-31 [1] CRAN (R 4.2.0)
#>  vctrs         0.6.2    2023-04-19 [1] CRAN (R 4.2.0)
#>  withr         2.5.0    2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun          0.38     2023-03-24 [1] CRAN (R 4.2.0)
#>  yaml          2.3.5    2022-02-21 [1] CRAN (R 4.2.0)
#>  zoo           1.8-12   2023-04-13 [1] CRAN (R 4.2.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

<sup>Created on 2023-05-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by gregmacfarlane** [comment]
The issue stems from [these lines](https://github.com/tidymodels/broom/blob/48099f0158520e44644284e561a9ceb4173aabd6/R/mlogit-tidiers.R#L86C2-L88) which presume that the indices always have the default names. 

```r
    dplyr::mutate(
      id = idx$id1,
      alternative = idx$id2,
```

PR #1156 has a proposed solution.

**Comment by simonpcouch** [comment]
Closed via #1156.
